src/app/components/login/login.component.ts

```diff
@@ -1,13 +1,64 @@
 import { Component } from '@angular/core';
+import {
+  FormControl,
+  FormGroupDirective,
+  NgForm,
+  Validators,
+} from '@angular/forms';
+import { ErrorStateMatcher } from '@angular/material/core';
 import { fromByteArray } from 'base64-js';
 
+export class MyErrorStateMatcher implements ErrorStateMatcher {
+  isErrorState(
+    control: FormControl | null,
+    form: FormGroupDirective | NgForm | null
+  ): boolean {
+    const isSubmitted = form && form.submitted;
+    return !!(
+      control &&
+      control.invalid &&
+      (control.dirty || control.touched || isSubmitted)
+    );
```

Comment:
```suggestion
    return 
      control &&
      control.invalid &&
      (control.dirty || control.touched || isSubmitted)
    ;
```

Once u have boolean operators, u do not need to have the casting value (!!)

Reply:
Tengo que ponerlo así porque control es formcontrol o nulo.